### DISCUSSION TITLE
Custom Field Uploadcare Image Uploads

### DISCUSSION THREAD

**Discussion by madsem** [filamentphp/filament#3074] [Help] [answered]
I'm trying myself on creating a custom field to implement Uploadcare image/video uploads:
https://uploadcare.com/docs/uploads/file-uploader/?_ga=2.111911431.2083330102.1657353744-305093601.1657353744

In this case, the uploader will return a `uuid` somehow, and I would need to save that along with the form... But so far I have no clue how :)

PS: removed the other part of the question, I realized that was actually just a config issue.

**Comment by danharrin** [comment, marked as answer]
Used Alpine before?

In the "custom fields" docs, see the 'state' Alpine property? Anything that gets saved to that property will be available in your form data. And then you load the view, the 'state' property will contain your DB information to populate the input with.

**Reply by madsem** [reply]
Yes, just not feeling super secure with it yet :)
Thank you @danharrin, that sounds pretty straightforward.

**Reply by madsem** [reply]
@danharrin  it sounded very easy, but after trying for 3 hours now to save values in the custom field, I concur that something else is up that goes beyond my current alpinejs knowledge :)

This is what I've tried:
```html
<x-forms::field-wrapper
    :id="$getId()"
    :label="$getLabel()"
    :label-sr-only="$isLabelHidden()"
    :helper-text="$getHelperText()"
    :hint="$getHint()"
    :hint-icon="$getHintIcon()"
    :required="$isRequired()"
    :state-path="$getStatePath()"
>
    <div x-data="{ state: $wire.entangle('{{ $getStatePath() }}') }" class="uploadcare-preview" wire:ignore>
        <input
            x-model="state"
            class="img-widget"
            type="hidden"
            role="uploadcare-uploader"
            data-public-key="{{ config('services.uploadcare.key') }}"
            data-tabs="all"
            data-preview-step=""
            data-clearable="true"
            data-images-only="true"
            data-effects="crop, rotate, mirror, sharp, blur, flip, enhance, grayscale, invert"
        />
        <div class="uploadcare-preview-image"></div>
    </div>
</x-forms::field-wrapper>
```

I've also tried removing the `$wire.entangle`,, set `state` to an empty string and just use `x-model` or `x-data`, nothing works.

The value attribute on the input, is set dynamically by Uploadcare after an image was uploaded.
Is that an issue for Alpine & custom fields?

**Reply by danharrin** [reply]
It looks like you're doing it right, have you gotten it working with just a simple text input, bound to `state` using `x-model`?

**Reply by madsem** [reply]
hey @danharrin great timing haha, I JUST now got it to work.
But it feels wonky...

```html
<x-forms::field-wrapper
    :id="$getId()"
    :label="$getLabel()"
    :label-sr-only="$isLabelHidden()"
    :helper-text="$getHelperText()"
    :hint="$getHint()"
    :hint-icon="$getHintIcon()"
    :required="$isRequired()"
    :state-path="$getStatePath()"
>
    <div x-data="{ state: $wire.entangle('{{ $getStatePath() }}') }" wire:ignore>
        <input
            x-model="state"
            x-ref="widget"
            x-init="await getImageUrl"
            class="img-widget"
            type="hidden"
            role="uploadcare-uploader"
            data-public-key="{{ config('services.uploadcare.key') }}"
            data-tabs="all"
            data-preview-step=""
            data-clearable="true"
            data-images-only="true"
            data-effects="crop, rotate, mirror, sharp, blur, flip, enhance, grayscale, invert"
        />
        <script>
            async function getImageUrl() {

                const uploader = window.uploadcare.SingleWidget(this.$refs.widget);

                // listen to the "upload completed" event
                uploader.onUploadComplete(fileInfo => {
                    // get a CDN URL from the file info
                    this.state = fileInfo.cdnUrl;
                });

            }
        </script>
    </div>

</x-forms::field-wrapper>
```

This works, and the cdn url is saved to DB.
But for some reason, I had to make the function async and add an `await` to `x-init`, without that the `window.uploadcare` is not defined...

This is the part that feels wonky, because really the async await isn't doing anything. I suppose that is a JS internal thing I don't know, that async functions are somehow fired after the `load` event?

Because simply wrapping the code inside `getImageUrl()` in an event listener for `load`, doesn't work.

**Reply by danharrin** [reply]
Looks good! Maybe release it as a plugin?

**Reply by madsem** [reply]
Absolutely, have to check how this is done :)

**Reply by michaelklopf** [reply]
@danharrin Any idea why this is working with await/async? It feels underwhelming that it works and I don't really understand why.
